File: CPP/DP/55. Jump Game prac.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int curEnd = 0, curReach = 0, goal = nums.size()-1;
        for (int i = 0; i <= goal; i++){
            curReach = max(curReach, i + nums[i]);
            if (curReach >= goal) return true;
            if (curReach == i && nums[i] == 0) return false;
        }
        return true;
    }
};

File: Python/DP/55. Jump Game prac.py

class Solution(object):
    def canJump(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        nums_len = len(nums)
        if nums_len == 1: return True
        MaxReach = 0
        for i in range(nums_len):
            CanReach = i + nums[i]
            if CanReach > MaxReach:
                MaxReach = CanReach
                if MaxReach >= nums_len - 1:
                    return True 
            if CanReach == i and MaxReach <= i:
                return False 

        # return False
